### PR TITLE
Rkyv persistence

### PR DESCRIPTION
**Pull request by krl**
This deprecates `canonical` in favor of rkyv, and implements a new strategy for storing data.

Resolves #76, #79

### CODE REVIEW

File: src/link.rs [hunk 1]
```diff
@@ -5,361 +5,216 @@
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
 use alloc::rc::Rc;
-use core::cell::{Ref, RefCell, RefMut};
-use core::mem;
-use core::ops::{Deref, DerefMut};
+use core::cell::RefCell;
+use rkyv::ser::Serializer;
+use std::borrow::BorrowMut;
```
[the remark this answers is not in the document]
Comment (krl) [reply src/link.rs]: fixed

File: src/storage/host.rs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+impl Default for Storage {
+    fn default() -> Self {
+        Storage {
+            lanes: Default::default(),
+            written: 0,
+        }
+    }
+}
+
+unsafe impl Sync for Storage {}
```
[the remark this answers is not in the document]
Comment (krl) [reply src/storage/host.rs]: this was left over from a previous refactor, i could just remove it.

File: src/storage/mod.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+use core::marker::PhantomData;
+
+use rkyv::{ser::Serializer, Archive, Fallible, Infallible};
+
+pub struct Stored<T> {
+    offset: u64,
+    _marker: PhantomData<T>,
+}
+
+unsafe impl<T> Send for Stored<T> {}
```
[the remark this answers is not in the document]
Comment (krl) [reply src/storage/mod.rs:19]: motivation added.

File: src/storage/guest.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+use super::Stored;
+
+use super::Portal;
+
+/// A representation of the current storage state
+pub struct Storage;
+
+impl Serializer for Storage {
+    fn pos(&self) -> usize {
+        todo!()
```
[the remark this answers is not in the document]
Comment (krl) [reply src/storage/guest.rs:18]: This is for the FFI part that is not implemented yet.

**Review by ureeves — APPROVED** [review]

**Pull request merged**